Functionize all of crm_shadow.c:main()

nrwahl2 opened the pull request:
One last step before formatted output. This is tedious but it ought to be done, and now is as good a time as any.

9e8a059 made the `crm_shadow.c` code somewhat easier to follow, but it's still a mess. The `main()` function is monolithic. Code is shared where possible -- often several commands share a particular initialization or validation step. This is typically handled by switch statements. However, it quickly becomes difficult to follow what any particular command is doing from start to finish. This makes it hard to reason about whether changes preserve existing behavior or not.

This PR improves `crm_shadow.c` by functionizing the logic of each command, after global option parsing. There is some duplication among the new functions, but the ease of maintenance justifies it.

nrwahl2 commented:
hoping to have something early (relative to my schedule) today (Thursday)

nrwahl2 commented:
I think this is done... I haven't had time to do a full self-review of the whole PR, but I'm on PTO all next week, so I wanted to get this pushed.

nrwahl2 commented:
rebased on current main and reworded a couple of comments

nrwahl2 commented on cts/cts-cli.in:
```diff
@@ -1903,7 +1899,7 @@ function test_tools() {
     desc="Reset shadow instance (mismatch)"
     cmd="crm_shadow --reset $shadow --batch"
-    test_assert $CRM_EX_USAGE 0
+    test_assert $CRM_EX_OK 0
```
On second thought, you could make a good case for still requiring `--force` for an instance mismatch in reset. Since it typically uses an already existing shadow file, you could assume that the user wants to reset the currently active one.

kgaillot replied:
Yeah I think it does make sense to keep --force for reset, just to make sure they have to think about which one they want

nrwahl2 commented on tools/crm_shadow.c:
```diff
@@ -252,18 +290,6 @@ shadow_teardown(char *name)
     free(our_prompt);
 }
 
-static bool
```
Despite the commit message saying these all fit nicely together, I might split some of these changes up on the next update. It might have gotten more complex since I wrote that.

kgaillot commented on tools/crm_shadow.c:
```diff
@@ -294,6 +294,82 @@ show_shadow_contents(GError **error)
+        goto done;
+    }
+
+    new_config = filename2xml(filename);
```
Good time to add a null check here

kgaillot reviewed (commented):
Reviewed everything through `Functionize "--switch"` before running out of time, looks great

kgaillot commented on tools/crm_shadow.c:
```diff
@@ -237,6 +237,33 @@ query_real_cib(xmlNode **output, GError **error)
+                    pcmk_rc_str(rc));
+        return rc;
+    }
+    return pcmk_rc_ok;
```
return rc works for both cases here

nrwahl2 replied:
see above

kgaillot reviewed (commented):
Looking good

nrwahl2 commented on tools/crm_shadow.c:
```diff
@@ -237,6 +237,33 @@ query_real_cib(xmlNode **output, GError **error)
+    int rc = write_xml_file(xml, filename, FALSE);
+
+    rc = pcmk_legacy2rc(rc);
+    if (rc < 0) {
```
Nope, I already made that mistake :) But I do need to move the `pcmk_legacy2rc()` inside the if block.

`write_xml_file()` returns the number of bytes written on success, -errno otherwise.

nrwahl2 commented:
Updated per review. The main change is that `--reset` is now treated as a dangerous command, such that it requires `--force` not only to override an instance mismatch but to run at all. This is consistent with `--delete`, which destroys a shadow file; `--reset` overwrites it, which is also destructive.

That is a change in behavior. If someone has a script that calls `crm_shadow --reset <instance>` without `--force`, it will break. So if we want to be really conservative, we can add a COMPAT note and keep the code a bit inconsistent for now. It seems somewhat unlikely though... I don't know that `crm_shadow` is used much outside of CTS.

I also moved the `pcmk_legacy2rc()` for `write_shadow_file()` into the if block (https://github.com/ClusterLabs/pacemaker/pull/3032#discussion_r1127048034).

nrwahl2 commented:
> crm shell uses it extensively. Let's definitely preserve the existing behavior. I'll create a task to change it at a compatibility break.

No problem, updated. I left the part where we have `--reset --force` create the shadow file if it doesn't already exist, but removed the change that requires `--force` for `--reset` in general. Now it only requires `--force` if there's a mismatch, like it did before.

kgaillot reviewed (approved):
Thanks!
